### PR TITLE
Add start stop e2e test

### PR DESCRIPTION
**Pull request by piersy**
### Description

This PR adds a test that starts and stops nodes in a small network and verifies that the network can continue to function correctly throughout the process.

A quick overview of what the test does:

1. Start 4 nodes
2. Verify they can confirm a tx
3. Stop one node
4. Verify the remaining 3 can confirm a tx
5. Stop another node
6. Verify the remaining 2 cannot confirm a tx
7. Start the last stopped node
8. Verify that the previously un-confirmable tx can now be confirmed by the 3 running nodes.
9. Verify that the 3 running nodes can confirm a further tx
10. Start the remaining node
11. Verify that the netowrk can process another tx and that all sent txs have been processed by all nodes.

### Other changes

This PR makes the minSyncPeers, dialHistoryExpiration and InboundThrottleTime configurable, in order for this test to run/run quickly these values needed to be set and they were previously hardcoded.

This PR also ensures correct shutdown of the miner, worker and backend.Backend, and fixes a number of data races/ race conditions. Each fix is described in the associated commit messages, it's probably easier to review this PR commit by commit, since these changes otherwise look unrelated.

I think this PR would be best merged rater than 'squashed and merged' due to the large number of commits containing unrelated fixes.

#### Upstream PRs that came out of this work
* https://github.com/ethereum/go-ethereum/pull/23673
* https://github.com/ethereum/go-ethereum/pull/23674
* https://github.com/ethereum/go-ethereum/pull/23675

### Tested

I managed to run this test 200 times on my local machine with no failures.

### Related issues

#1642 

### Backwards compatibility

Yes

### CODE REVIEW

**Comment by piersy** [comment]
Coverage from tests in `./e2e_test/...` for `./consensus/istanbul/...` at commit d824233e0f014ca623350e72596955e9310a4d04
```
coverage:  50.2% of statements across all listed packages
coverage:  64.1% of statements in consensus/istanbul
coverage:  51.0% of statements in consensus/istanbul/backend
coverage:   0.0% of statements in consensus/istanbul/backend/backendtest
coverage:  45.0% of statements in consensus/istanbul/backend/internal/db
coverage:  42.5% of statements in consensus/istanbul/backend/internal/enodes
coverage:  23.6% of statements in consensus/istanbul/backend/internal/replica
coverage:  68.3% of statements in consensus/istanbul/core
coverage:   0.0% of statements in consensus/istanbul/proxy
coverage:  75.3% of statements in consensus/istanbul/uptime
coverage: 100.0% of statements in consensus/istanbul/uptime/store
coverage:  51.8% of statements in consensus/istanbul/validator
coverage:  79.2% of statements in consensus/istanbul/validator/random
```

**Comment by piersy** [comment]
Coverage from tests in `./e2e_test/...` for `./consensus/istanbul/...` at commit ef387af92745fc9c12f831e41199f0a853455c07
```
coverage:  46.8% of statements across all listed packages
coverage:  55.7% of statements in consensus/istanbul
coverage:  46.3% of statements in consensus/istanbul/backend
coverage:   0.0% of statements in consensus/istanbul/backend/backendtest
coverage:  45.0% of statements in consensus/istanbul/backend/internal/db
coverage:  28.9% of statements in consensus/istanbul/backend/internal/enodes
coverage:  23.6% of statements in consensus/istanbul/backend/internal/replica
coverage:  68.2% of statements in consensus/istanbul/core
coverage:   0.0% of statements in consensus/istanbul/proxy
coverage:  75.3% of statements in consensus/istanbul/uptime
coverage: 100.0% of statements in consensus/istanbul/uptime/store
coverage:  51.8% of statements in consensus/istanbul/validator
coverage:  79.2% of statements in consensus/istanbul/validator/random
```

**Comment by piersy** [comment]
Coverage from tests in `./e2e_test/...` for `./consensus/istanbul/...` at commit 204d06570daa327178b45bd105a2df8f26f6e68e
```
coverage:  45.7% of statements across all listed packages
coverage:  55.4% of statements in consensus/istanbul
coverage:  46.3% of statements in consensus/istanbul/backend
coverage:   0.0% of statements in consensus/istanbul/backend/backendtest
coverage:  45.0% of statements in consensus/istanbul/backend/internal/db
coverage:  28.9% of statements in consensus/istanbul/backend/internal/enodes
coverage:  23.6% of statements in consensus/istanbul/backend/internal/replica
coverage:  63.9% of statements in consensus/istanbul/core
coverage:   0.0% of statements in consensus/istanbul/proxy
coverage:  75.3% of statements in consensus/istanbul/uptime
coverage: 100.0% of statements in consensus/istanbul/uptime/store
coverage:  51.8% of statements in consensus/istanbul/validator
coverage:  79.2% of statements in consensus/istanbul/validator/random
```

**Comment by piersy** [comment]
Coverage from tests in `./e2e_test/...` for `./consensus/istanbul/...` at commit 155e71d0e9c1d2fc3e8f4f9c87b1c5687cbb400e
```
coverage:  50.0% of statements across all listed packages
coverage:  64.1% of statements in consensus/istanbul
coverage:  51.0% of statements in consensus/istanbul/backend
coverage:   0.0% of statements in consensus/istanbul/backend/backendtest
coverage:  45.0% of statements in consensus/istanbul/backend/internal/db
coverage:  42.5% of statements in consensus/istanbul/backend/internal/enodes
coverage:  23.6% of statements in consensus/istanbul/backend/internal/replica
coverage:  67.7% of statements in consensus/istanbul/core
coverage:   0.0% of statements in consensus/istanbul/proxy
coverage:  75.3% of statements in consensus/istanbul/uptime
coverage: 100.0% of statements in consensus/istanbul/uptime/store
coverage:  51.8% of statements in consensus/istanbul/validator
coverage:  79.2% of statements in consensus/istanbul/validator/random
```

**Comment by piersy** [comment]
Coverage from tests in `./e2e_test/...` for `./consensus/istanbul/...` at commit 00a9f7b1e1e73737ea15aa488e47a49c443f0e28
```
coverage:  50.9% of statements across all listed packages
coverage:  64.1% of statements in consensus/istanbul
coverage:  51.7% of statements in consensus/istanbul/backend
coverage:   0.0% of statements in consensus/istanbul/backend/backendtest
coverage:  50.0% of statements in consensus/istanbul/backend/internal/db
coverage:  46.5% of statements in consensus/istanbul/backend/internal/enodes
coverage:  23.6% of statements in consensus/istanbul/backend/internal/replica
coverage:  68.9% of statements in consensus/istanbul/core
coverage:   0.0% of statements in consensus/istanbul/proxy
coverage:  75.3% of statements in consensus/istanbul/uptime
coverage: 100.0% of statements in consensus/istanbul/uptime/store
coverage:  51.8% of statements in consensus/istanbul/validator
coverage:  79.2% of statements in consensus/istanbul/validator/random
```

**Comment by piersy** [comment]
Coverage from tests in `./e2e_test/...` for `./consensus/istanbul/...` at commit 00358de553049fd32ff06c261f0ff9f69857eeaa
```
coverage:  51.4% of statements across all listed packages
coverage:  64.1% of statements in consensus/istanbul
coverage:  52.1% of statements in consensus/istanbul/backend
coverage:   0.0% of statements in consensus/istanbul/backend/backendtest
coverage:  50.0% of statements in consensus/istanbul/backend/internal/db
coverage:  46.0% of statements in consensus/istanbul/backend/internal/enodes
coverage:  23.6% of statements in consensus/istanbul/backend/internal/replica
coverage:  70.3% of statements in consensus/istanbul/core
coverage:   0.0% of statements in consensus/istanbul/proxy
coverage:  75.3% of statements in consensus/istanbul/uptime
coverage: 100.0% of statements in consensus/istanbul/uptime/store
coverage:  51.8% of statements in consensus/istanbul/validator
coverage:  79.2% of statements in consensus/istanbul/validator/random
```

**Comment by piersy** [comment]
Coverage from tests in `./e2e_test/...` for `./consensus/istanbul/...` at commit a04b02099a9a735fdc0bc4d897e3aa08b1f951f3
```
coverage:  51.5% of statements across all listed packages
coverage:  64.1% of statements in consensus/istanbul
coverage:  51.8% of statements in consensus/istanbul/backend
coverage:   0.0% of statements in consensus/istanbul/backend/backendtest
coverage:  50.0% of statements in consensus/istanbul/backend/internal/db
coverage:  46.5% of statements in consensus/istanbul/backend/internal/enodes
coverage:  23.6% of statements in consensus/istanbul/backend/internal/replica
coverage:  71.0% of statements in consensus/istanbul/core
coverage:   0.0% of statements in consensus/istanbul/proxy
coverage:  75.3% of statements in consensus/istanbul/uptime
coverage: 100.0% of statements in consensus/istanbul/uptime/store
coverage:  51.8% of statements in consensus/istanbul/validator
coverage:  79.2% of statements in consensus/istanbul/validator/random
```

**Comment by piersy** [comment]
Coverage from tests in `./e2e_test/...` for `./consensus/istanbul/...` at commit 99a2d7f8899132824522132b66804fb1f37e5b05
```
coverage:  46.9% of statements across all listed packages
coverage:  55.4% of statements in consensus/istanbul
coverage:  46.3% of statements in consensus/istanbul/backend
coverage:   0.0% of statements in consensus/istanbul/backend/backendtest
coverage:  45.0% of statements in consensus/istanbul/backend/internal/db
coverage:  28.9% of statements in consensus/istanbul/backend/internal/enodes
coverage:  23.6% of statements in consensus/istanbul/backend/internal/replica
coverage:  68.8% of statements in consensus/istanbul/core
coverage:   0.0% of statements in consensus/istanbul/proxy
coverage:  75.3% of statements in consensus/istanbul/uptime
coverage: 100.0% of statements in consensus/istanbul/uptime/store
coverage:  51.8% of statements in consensus/istanbul/validator
coverage:  79.2% of statements in consensus/istanbul/validator/random
```

File: consensus/istanbul/backend/backend.go [hunk 1]
```diff
@@ -210,6 +210,7 @@ type Backend struct {
 
 	aWallets atomic.Value
 
+	wg           sync.WaitGroup
```
Comment (hbandura) [inline consensus/istanbul/backend/backend.go:213]: Given the amount of properties backend already has, I'd consider adding a much more descriptive variable name for this waitgroup (and a small comment explaining why it may be needed). `closeReadyWG`, or something along those lines

File: consensus/istanbul/backend/handler.go [hunk 2]
```diff
@@ -346,6 +346,8 @@ func (sb *Backend) newChainHead(newBlock *types.Block) {
 
 		sb.logger.Info("Validator Election Results", "address", sb.ValidatorAddress(), "elected", valSetIndex >= 0, "number", newBlock.Number().Uint64())
 
+		sb.announceMu.Lock()
+		defer sb.announceMu.Unlock()
 		if sb.announceRunning {
```
Comment (hbandura) [inline consensus/istanbul/backend/handler.go:351]: Mind adding a comment on why is this necessary? Is it for protecting the access of the announceRunning variable, or because of the RefreshValPeers?

File: core/blockchain.go [hunk 3]
```diff
@@ -196,6 +196,7 @@ type BlockChain struct {
 
 	quit          chan struct{}  // blockchain quit channel
 	wg            sync.WaitGroup // chain processing wait group for shutting down
+	wgMu          sync.Mutex     // chain processing wait group for shutting down
```
Comment (hbandura) [inline core/blockchain.go:199]: Is this mutex intended to make sure the `Wait` happens after the first `Add`? AFAIK `sync.WaitGroup` is thread safe in its use and there shouldn't be any need to syncronize its access, but I may be missing some border case

File: trie/database.go [hunk 4]
```diff
@@ -778,10 +778,15 @@ func (db *Database) Commit(node common.Hash, report bool, callback func(common.H
 // commit is the private locked version of Commit.
 func (db *Database) commit(hash common.Hash, batch ethdb.Batch, uncacher *cleaner, callback func(common.Hash)) error {
 	// If the node does not exist, it's a previously committed node
+
+	db.lock.RLock()
 	node, ok := db.dirties[hash]
 	if !ok {
```
Comment (hbandura) [inline trie/database.go:784]: Isn't this the same as moving the `RUnlock` call right after the `db.dirties[hash]` read?

**Comment by hbandura** [comment]
I'm confused about some changes. Are the `p2p` changes and the `eth/config` changes actually part of this PR or is github botching the diff due to a merge? If they are, they seem like big (well, not big, but important) and I'd rather have them in a separate PR to make them more standout more

File: core/blockchain.go (hunk 3 above)
Comment (piersy) [reply core/blockchain.go:199]: This mutex ensure that wait and add cannot be called concurrently see - https://github.com/golang/go/issues/23842

File: trie/database.go (hunk 4 above)
Comment (piersy) [reply trie/database.go:784]: Yes it is :)  thank you !

**Comment by piersy** [comment]
Coverage from tests in `./e2e_test/...` for `./consensus/istanbul/...` at commit ff49fdfe77a7ae5afee550a60425efc4fda5d22a
```
coverage:  49.9% of statements across all listed packages
coverage:  64.1% of statements in consensus/istanbul
coverage:  50.9% of statements in consensus/istanbul/backend
coverage:   0.0% of statements in consensus/istanbul/backend/backendtest
coverage:  45.0% of statements in consensus/istanbul/backend/internal/db
coverage:  42.8% of statements in consensus/istanbul/backend/internal/enodes
coverage:  23.6% of statements in consensus/istanbul/backend/internal/replica
coverage:  67.5% of statements in consensus/istanbul/core
coverage:   0.0% of statements in consensus/istanbul/proxy
coverage:  75.3% of statements in consensus/istanbul/uptime
coverage: 100.0% of statements in consensus/istanbul/uptime/store
coverage:  51.8% of statements in consensus/istanbul/validator
coverage:  79.2% of statements in consensus/istanbul/validator/random
```

File: consensus/istanbul/backend/handler.go (hunk 2 above)
Comment (piersy) [reply consensus/istanbul/backend/handler.go:354]: Done

**Comment by piersy** [comment]
Coverage from tests in `./e2e_test/...` for `./consensus/istanbul/...` at commit 20cc682b9acfc749c2967fbfdc121167c143c9fa
```
coverage:  50.9% of statements across all listed packages
coverage:  63.6% of statements in consensus/istanbul
coverage:  51.9% of statements in consensus/istanbul/backend
coverage:   0.0% of statements in consensus/istanbul/backend/backendtest
coverage:  50.0% of statements in consensus/istanbul/backend/internal/db
coverage:  46.3% of statements in consensus/istanbul/backend/internal/enodes
coverage:  23.6% of statements in consensus/istanbul/backend/internal/replica
coverage:  68.7% of statements in consensus/istanbul/core
coverage:   0.0% of statements in consensus/istanbul/proxy
coverage:  75.3% of statements in consensus/istanbul/uptime
coverage: 100.0% of statements in consensus/istanbul/uptime/store
coverage:  51.8% of statements in consensus/istanbul/validator
coverage:  79.2% of statements in consensus/istanbul/validator/random
```

**Comment by piersy** [comment]
> I'm confused about some changes. Are the `p2p` changes and the `eth/config` changes actually part of this PR or is github botching the diff due to a merge? If they are, they seem like big (well, not big, but important) and I'd rather have them in a separate PR to make them more standout more

Hey @hbandura they are meant to be part of this pr. I think the commit comments make this clear.

> Allow configuring the min peers needed to sync
In order to run small scale e2e tests reliably we need to be able to let
nodes sync from just one other node. This commit allows the minimum sync
peers to be configured, where previously it was hardcoded to 5.

And 

>Allow modification of dial and connection timeouts
Now dialHistoryExpiration and inboundThrottleTime can be configured via
config. This allows for the writing of fast tests.

I would suggest just checking this PR commit by commit for a better understanding of all the individual changes that it contains.

We could still split these changes out, if you think that would be better. I generally try to keep changes with the thing that precipitated the changes, since that makes it clear why the changes are needed. Otherwise you end up with PRs having changes that don't seem to serve any purpose. That's why I kept all these changes together because they are all needed for this test to work.

**Comment by piersy** [comment]
Coverage from tests in `./e2e_test/...` for `./consensus/istanbul/...` at commit b390d1f545d23d5022280428c991a1c46ca8e302
<details><summary>coverage:  46.5% of statements across all listed packages</summary><br>
<pre>
coverage:  55.4% of statements in consensus/istanbul
coverage:  46.3% of statements in consensus/istanbul/backend
coverage:   0.0% of statements in consensus/istanbul/backend/backendtest
coverage:  45.0% of statements in consensus/istanbul/backend/internal/db
coverage:  28.9% of statements in consensus/istanbul/backend/internal/enodes
coverage:  23.6% of statements in consensus/istanbul/backend/internal/replica
coverage:  67.1% of statements in consensus/istanbul/core
coverage:   0.0% of statements in consensus/istanbul/proxy
coverage:  75.3% of statements in consensus/istanbul/uptime
coverage: 100.0% of statements in consensus/istanbul/uptime/store
coverage:  51.8% of statements in consensus/istanbul/validator
coverage:  79.2% of statements in consensus/istanbul/validator/random
</pre>
CommentID: 9bd7c06e8e
</details>

File: core/blockchain.go (hunk 3 above)
Comment (hbandura) [reply core/blockchain.go:199]: Ah, right, it's thread safe if you call Add when the count is already > 0, but if it's exactly 0 then it's not thread safe. Either a broken API or a weird use by ourselves.

Given that in this particular use case Add should never be called from a different goroutine without the other lock, I think we better encapsulate this in a new type (`DynamicWaitGroup` or whatever name you can think of) and have its own internal mutex, to avoid adding one more reference to the blockchain type, and also making sure that any new code written here cannot accidentally call the Add without the lock.

File: consensus/istanbul/backend/backend.go [hunk 5]
```diff
@@ -549,7 +551,11 @@ func (sb *Backend) Commit(proposal istanbul.Proposal, aggregatedSeal types.Istan
 			return err
 		}
 	}
-	go sb.onNewConsensusBlock(block, result.Receipts, result.Logs, result.State)
+	sb.wg.Add(1)
```
Comment (hbandura) [inline consensus/istanbul/backend/backend.go:554]: Doesn't this waitgroup face the same problem with the concurrent `Add` ? or are Commit and Close both called from the same thread?

File: consensus/istanbul/backend/backend.go (hunk 1 above)
Comment (piersy) [reply consensus/istanbul/backend/backend.go]: So it looks like we may just be able to remove this wg, since the go-routine it was waiting for doesn't appear to be necessary.

File: internal/syncx/mutex.go [hunk 6]
```diff
@@ -0,0 +1,64 @@
+// Copyright 2021 The go-ethereum Authors
```
Comment (hbandura) [inline internal/syncx/mutex.go:1]: "The Celo Authors"

File: consensus/istanbul/backend/backend.go [hunk 7]
```diff
@@ -549,8 +549,7 @@ func (sb *Backend) Commit(proposal istanbul.Proposal, aggregatedSeal types.Istan
 			return err
 		}
 	}
-	go sb.onNewConsensusBlock(block, result.Receipts, result.Logs, result.State)
-
+	sb.onNewConsensusBlock(block, result.Receipts, result.Logs, result.State)
```
Comment (hbandura) [inline consensus/istanbul/backend/backend.go:552]: Is running this on a goroutine no longer necessary due to the other syncronization added in worker?

File: miner/worker.go [hunk 8]
```diff
@@ -351,6 +357,8 @@ func (w *worker) mainLoop() {
 	var cancel context.CancelFunc
 	var wg sync.WaitGroup
 
+	// Ensure that block construction is complete before exiting this function.
```
Comment (hbandura) [inline miner/worker.go:360]: expand on why block construction is needed before exiting, if possible
